lib/ABI/FunctionType/Layout.cpp

```diff
@@ -437,32 +428,29 @@ Layout::Layout(const model::CABIFunctionType &Function) {
                                                  RV.SizeOnStack != 0);
   revng_assert(Converted.size() >= Function.Arguments().size());
   uint64_t Index = 0;
+  uint64_t StackStructSize = 0;
   for (const DistributedValue &Distributed : Converted) {
     if (!Distributed.RepresentsPadding) {
       Argument &Current = Arguments.emplace_back();
-      const auto &ArgumentType = Function.Arguments().at(Index).Type();
+      Current.Type = Function.Arguments().at(Index).Type();
+      if (Distributed.UsesPointerToCopy == true)
+        Current.Type = Current.Type.getPointerTo(ABI.getArchitecture());
 
       // Disambiguate scalar and aggregate arguments.
       // Scalars are passed by value, aggregates - by pointer.
-      Current.Type = ArgumentType;
-      if (ArgumentType.isScalar())
+      if (Current.Type.isScalar())
         Current.Kind = ArgumentKind::Scalar;
       else
         Current.Kind = ArgumentKind::ReferenceToAggregate;
 
       Current.Registers = std::move(Distributed.Registers);
       if (Distributed.SizeOnStack != 0) {
         // The argument has a part (or is placed entirely) on the stack.
-        Current.Stack = Layout::Argument::StackSpan{};
-
-        // Round the offset based on the natural alignment,
-        CurrentStackOffset = ABI.alignedOffset(CurrentStackOffset,
-                                               ArgumentType);
-        Current.Stack->Offset = CurrentStackOffset;
+        Current.Stack = Layout::Argument::StackSpan{
+          .Offset = Distributed.OffsetOnStack, .Size = Distributed.SizeOnStack
+        };
 
-        // And carry the size over unchanged.
-        Current.Stack->Size = Distributed.SizeOnStack;
-        CurrentStackOffset += Current.Stack->Size;
+        StackStructSize = Distributed.OffsetOnStack + Distributed.SizeOnStack;
```

Comment:
> `ToRaw: take advantage of the updated distributor`: what changes specifically? Consider squashing commits.

@aleclearmind, specifically, it drops all the alignment-related computations that used to be done "at the leaves" and replaces them with data that's newly embedded into the `ArgumentDistributor` in of the previous commits.

Reply:
OK, can you elaborate on this a bit in the commit message?

Reply:
Done:
```
ToRaw: take advantage of the updated distributor

This lets us drop the repeated alignment calculations at
the "leaves" of the conversion. Since new distributor already
provides the necessary information, we can just grab it from
there.
```